### PR TITLE
All Enhancement proposals in single webpage | KEPs

### PR DESCRIPTION
**Pull request by shekhar-rajak**
This PR is for https://github.com/kubernetes/enhancements/issues/2095 and we want a KEP webpage to find out about specific kep and its metadata easily.

1. keps.json is generated from enhancements repo using gen-content.sh
2. sigs.yaml is fetched from community/sigs.yaml file

Refer: https://github.com/kubernetes/website/pull/28295

### CODE REVIEW

**Comment by shekhar-rajak** [comment]
Please check the page: https://deploy-preview-222--kubernetes-contributor.netlify.app/keps/

File: layouts/shortcodes/keps-data.html
Comment (sftim) [inline layouts/shortcodes/keps-data.html:32]: Optional: with
```suggestion
      <tr id="kep-row-$data.kep_number">
```
then links to KEP rows become possible.

**Review by sftim — COMMENTED** [review]
Hi @shekhar-rajak 

Here's a much less formal review, with suggestions if you want to take them.

File: layouts/shortcodes/keps-data.html
Comment (mrbobbytables) [inline layouts/shortcodes/keps-data.html:33]: `https://kep.k8s.io/<kep_number>` (e.g. http://kep.k8s.io/1027) Will take you right to the tracking issue

File: layouts/shortcodes/keps-data.html
Comment (palnabarun) [inline layouts/shortcodes/keps-data.html:14]: We don't need to pull in the complete sigs.yaml to just get the name of the SIG. The KEP data already has all the SIGs in each KEP entry. We can think about this filter by SIG feature in a later iteration.

Note: The list of SIGs can also be populated by JavaScript at page load by iterating over the values in SIG column.

File: layouts/shortcodes/keps-data.html
Comment (palnabarun) [inline layouts/shortcodes/keps-data.html:11]:
```suggestion
        <label for="keps">Choose SIG:</label>
```

nit: redundancy

File: content/en/keps/_index.md
Comment (palnabarun) [inline content/en/keps/_index.md:12]: Possible rephrase.
```suggestion
> The data is generated from the [Enhancements repository](https://git.k8s.io/enhancements).
```

File: layouts/shortcodes/keps-data.html
Comment (palnabarun) [inline layouts/shortcodes/keps-data.html:37]: Since authors is an array, we should iterate over the entries and concatenate them separated by a comma.

Right now, the render looks like this:
```
[@foo @bar @baz]
```

The following would look better:
```
@foo, @bar, @baz
```


Possible improvement: The author handles are their GitHub usernames. We can even link to their account.

**Review by palnabarun — CHANGES_REQUESTED** [review]

File: layouts/shortcodes/keps-data.html
Comment (palnabarun) [reply layouts/shortcodes/keps-data.html:36]: +1

File: layouts/shortcodes/keps-data.html
Comment (palnabarun) [reply layouts/shortcodes/keps-data.html:30]: The sort order right now is weird. I think it's unsorted. Ideally, it should be ordered by the KEP number in ascending order.

Comment (sftim) [reply layouts/shortcodes/keps-data.html:30]: :+1: Ordering by KEP number seems good for an initial release.

File: layouts/shortcodes/keps-data.html
Comment (sftim) [reply layouts/shortcodes/keps-data.html:37]: I'd be happy to save both those improvements for later. They're definitely nice things to have.

File: static/js/keps/kep_search.js
Comment (shekhar-rajak) [reply static/js/keps/kep_search.js:5]: Thanks for review. Once we type something in search box or paste any content it will filter the table - we do not need submit button.

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:31]: Yeah, That needs to be worked on enhancement repo. I think we should update those KEPs metadata there.

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:30]: Good point. It is better to have option to sort keps while generating the YAML data using `kepctl query` API. For now I will try it using jquery.

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:36]: I felt that it is better to display the name of the SIG.
I will update the link part.

File: layouts/shortcodes/keps-data.html
Comment (sftim) [reply layouts/shortcodes/keps-data.html:30]: You can also use https://gohugo.io/functions/sort/

eg
```go
{{ range $index, $element := sort $.Site.Data.keps.keps_data.keps "kep_number" }}
```

(bear in mind that `kep_number` is a string and that this might not sort how you hoped)

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:14]: Yeah, I wanted to have options to choose complete list of SIGs we have. So lets say `sig-usability` keps we do not have in enhancements repo, but we still like to show it in dropdown.

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:37]: Yup, I thought about it. I will try for it.

**Comment by shekhar-rajak** [comment]
> Instead of importing directly with the hack script we can use [data template to call a url](https://gohugo.io/templates/data-templates/#call-the-functions-with-a-url) and grab an updated copy of the data. A postsubmit job in the enhancements repo that generates it and dumps it to a gcs bucket with an always up to date copy would be super.

Thanks for the suggestion. Yes, we need to have CI script that will generate the YAML/JSON file if the merge has been done for new KEP and upload it. I would like to this TODO to be completed once this page is ready and merged. 

If you have suggestion how to upload in gcs bucket , please share any link/documentations/sample PR.

File: layouts/shortcodes/keps-data.html
Comment (shekhar-rajak) [reply layouts/shortcodes/keps-data.html:30]: Thanks!

**Comment by sftim** [comment]
https://deploy-preview-222--kubernetes-contributor.netlify.app/keps/ is looking pretty good.

(IMO) we should only merge this if we're happy with the toil to keep it current, though. If not, let's find a way to help get that JSON document with the KEP and SIG data available so it Hugo can fetch it via HTTP GET.

(One option that springs to mind: a GitHub Pages site).

**Comment by shekhar-rajak** [comment]
A GitHub bot need to upload it either in Gist/repo or any storage bucket whenever a KEP is merged in enhancement repo similarly any changes in sigs.yaml file : https://raw.githubusercontent.com/kubernetes/community/master/sigs.yaml

File: layouts/shortcodes/keps-data.html
Comment (sftim) [inline layouts/shortcodes/keps-data.html:54]: This normalizes the hyperlink and GitHub username (and makes styling a bit easier):
```suggestion
            <a href="https://github.com/{{ trim $author "@" }}">
              <span class="github-user">@{{ trim $author "@" }}</span>
```

File: layouts/shortcodes/keps-data.html
Comment (sftim) [inline layouts/shortcodes/keps-data.html:37]: Future enhancement: replace incorrect KEP numbers with a styled `?`.

File: static/js/keps/kep_search.js
Comment (shekhar-rajak) [reply static/js/keps/kep_search.js:5]: Yeah, I got that. But when I tried pasting any string in input box it was filtering . Please check once and let me know if you find issue

**Comment by sftim** [comment]
The main thing I think this needs is automation. I'd be reluctant to see this merge unless either:
- someone commits to keep this up to date manually, as often as is needed
- we instead have some code to do that updating and are happy that it'll work

**Comment by shekhar-rajak** [comment]
Yeah @sftim , automation is needed. I want to use a URL. instead of having keps_data.yaml file in data folder, so that it will fetch from there. We can have something like this : https://github.com/kubernetes/community/blob/master/sigs.yaml so that we can get it from raw github userdata URL: https://github.com/kubernetes/community/raw/master/sigs.yaml

I have started the discussions here: 

* https://github.com/kubernetes/enhancements/pull/2801 
* https://github.com/kubernetes/community/issues/5862

Since Hugo doesn't suppot getYAML we only have getJSON so I want to have sig metadata in JSON

**Comment by shekhar-rajak** [comment]
I have explored hack/gen-content.sh and now this PR do not have any dependency. 

1. It is fetching the sigs.yaml from the community repo
2. It is generating keps.json file from the enhancements repo - kepctl query command

Only one change we have to do : https://github.com/kubernetes/enhancements/pull/2801 
kep-number should be kep_number otherwise Hugo do not understand the key name in json

**Comment by shekhar-rajak** [comment]
Preview available: https://deploy-preview-222--kubernetes-contributor.netlify.app/keps/

**Comment by sftim** [comment]
@shekhar-rajak the sort of automation I'm thinking of is a scheduled GitHub Action to build the KEP list page automatically so that if [upstream](https://github.com/kubernetes/enhancements/) changes, this site is then updated within a few hours.

(Triggered updates would be a nice-to-have but can wait for a long time - few people are looking for instant updates in a page like this).

**Comment by shekhar-rajak** [comment]
Thanks @sftim , yes we have to have such automation:

- Either we have trigger the build job if enhancement repo is updated 
- Or we have to put the keps.json file in some storage if any PR is merged in enhancement related to KEP change.

**Comment by shekhar-rajak** [comment]
Hi,

I believe this PR is complete for now. Once we get to know solution for AWS S3 storage or GCP storage to put the KEPs JSON data, we will just need to put the URL instead of the file path.  So this PR can be merged with TODOs and TODOs can be done in next PR.

Please let me know which team can help in providing the storage.

**Comment by sftim** [comment]
The team to ask about storage is, _I think_, WG Kubernetes Infrastructure - #wg-k8s-infra on Kubernetes Slack.

File: content/en/keps/_index.md
Comment (mrbobbytables) [inline content/en/keps/_index.md:1]: keps should not be a top level item. I think it'd make more sense under resources.

**Review by mrbobbytables — CHANGES_REQUESTED** [review]
Sorry for the delay in reviews, it dropped through the cracks for me.
I have 2 suggestions that might require some modification.

**Comment by shekhar-rajak** [comment]
Starting the slack thread https://kubernetes.slack.com/archives/CCK68P2Q2/p1628494118056300 for storage

File: hack/gen-content.sh
Comment (shekhar-rajak) [reply hack/gen-content.sh:31]: Yeah this will be removed once I get solution to put the meta data file in some storage object.

File: content/en/keps/_index.md
Comment (shekhar-rajak) [reply content/en/keps/_index.md:1]: I will work on it.

**Comment by spiffxp** [comment]
To briefly repost from https://github.com/kubernetes/k8s.io/issues/2490#issuecomment-896318679

I'm not saying no, but simply asking are you _sure_ you want a GCS bucket for this?  Multiple sources of truth tend to drift.  I'm a much bigger fan of trying to keep source of truth in a single location (repo) if it can be helped.  I suggested a few alternatives.

**Comment by kikisdeliveryservice** [comment]
> To briefly repost from [kubernetes/k8s.io#2490 (comment)](https://github.com/kubernetes/k8s.io/issues/2490#issuecomment-896318679)
> 
> I'm not saying no, but simply asking are you _sure_ you want a GCS bucket for this? Multiple sources of truth tend to drift. I'm a much bigger fan of trying to keep source of truth in a single location (repo) if it can be helped. I suggested a few alternatives.

I think @spiffxp proposed some good suggestions in this linked post above and think they are worth considering as I agree that it's preferable to maintain a single source of truth as opposed to a repo and bucket.

**Comment by spiffxp** [comment]
Poke.  If it's a bucket you want, I have https://github.com/kubernetes/k8s.io/pull/2570 up for consideration, PTAL

**Comment by sftim** [comment]
Hi. When I imagined this with a bucket, the job of keeping its contents current was going to fall to automation that builds based on what's in [k/enhancements](https://github.com/kubernetes/enhancements/).

That approach plus this PR plus a bit more automation in this repo _ought_ to entail no extra effort for folks proposing an enhancement, but better rendering (automatically) when they do, and possibly also better short links such as https://kep.k8s.io/2225 (not yet working)

**Comment by spiffxp** [comment]
Poke again.

>When I imagined this with a bucket, the job of keeping its contents current was going to fall to automation that builds based on what's in k/enhancements.

https://github.com/kubernetes/k8s.io/pull/2570 is written with that perspective in mind.  Does that set of infra meet our needs?

Where we are blocked is deciding who the appropriate humans are for purposes of ownership for troubleshooting, support, and escalation.  It's k8s-infra's job to make sure that the appropriate set of infrastructure is present and that the right group of people can access it.  But we're not going to provision it if there are no humans willing to support it.

**Comment by spiffxp** [comment]
Rebased https://github.com/kubernetes/k8s.io/pull/2570, arbitrarily chose `enhancements-approvers` to own the infrastructure, please raise objections there.

And please confirm it's what you want and you're going to use it.  Last thing we want is zombie infrastructure.  I'm going to close due to lack of activity before too long.

**Comment by shekhar-rajak** [comment]
As far as I understand the usecase

1. Write access the latest keps.yaml generate by each PR in GitHub repo kubernetes/enhancements  to the gap bucket 
2. Read access to the account in this repo kubernetes/contributor-site
 
I would like to know , which account will be having read/write access to the bucket when CI/CD runs.
Please share If there is any sample PR/template that I can refer.

**Comment by sftim** [comment]
@salaxander you previously suggested you might have time to help out with parts of this improvement. Does this PR look like a part you'd have time to help with?

**Comment by shekhar-rajak** [comment]
The PR https://github.com/kubernetes/k8s.io/pull/2570 is merged. I will continue work on this PR.

**Comment by sftim** [comment]
How's this going? What help would you like, @shekhar-rajak ?

**Comment by shekhar-rajak** [comment]
Hi @sftim ,

Please share a sample PR that uses GCP bucket for read and write to a file ? 

* I want to upload file in GCP bucket automatically once any PR is merged into repo kubernetes/enhancements (later on we will improve it to trigger it when specific file is updated) - if there is bot or any CI/CD is integrated then please let me know

* Repo kubernetes/contributor-site will use the uploaded file to fetch the YAML/JSON data to display in website.

**Comment by sftim** [comment]
> share a sample PR that uses GCP bucket for read and write to a file ?

The read should be easy ([HTTP fetch](https://cloud.google.com/storage/docs/hosting-static-website-http) and get Hugo to load that JSON rather than locally from `data/keps/*`). The write, I'm not sure how we usually do this.

**Comment by sftim** [comment]
CI/CD for Kubernetes projects is something I don't know much about I'm afraid. I can try to find people who do.

**Comment by PushkarJ** [comment]
Hi @shekhar-rajak , I came across this issue (thanks to @sftim) while exploring something similar to what you are doing (writing to a bucket). I don't have a PR yet for the work I am doing, but found some useful resources and prior art that might give you hints on how to write what you want to write to the Google Cloud Bucket. 

- So general way to read/write to GCB is using `gsutil cp` command. More in docs here: https://cloud.google.com/storage/docs/quickstart-gsutil#upload_an_object_into_your_bucket

- One example where this is used with Prow is in this [shell script](https://github.com/kubernetes/test-infra/blob/8bddea4df208ea536b18e5eed7edb2328148f7e6/triage/update_summaries.sh#L99-L103), which is used in this [job](https://github.com/kubernetes/test-infra/blob/6dec48b89caab69806c17f72dd32d88ba1b1dcc0/config/jobs/kubernetes/sig-k8s-infra/trusted/sig-k8s-infra-test-infra.yaml#L84-L114)

Please ignore me if you have already figured this out :)

**Comment by shekhar-rajak** [comment]
Thank you @PushkarJ ! I will explore and try to use the similar existing way.

**Comment by sftim** [comment]
I'm no Prow expert - I'm not even a novice, really - but I'm happy to try to clarify any docs, provide pointers, interpret advice etc if it helps move this work forward. Feel free to ask questions @shekhar-rajak !

**Comment by sftim** [comment]
Still worth doing IMO.
/remove-lifecycle stale

**Comment by palnabarun** [comment]
/close

(in favor of #324)

**Pull request closed without merging**
